`seccomp_export_bpf_mem`

Upstream commit: https://github.com/seccomp/libseccomp/commit/3f0e47fe2717b73ccef68ca18f9f7297ee73ebb2 (not release yet)


`libseccomp-sys`:
```rust
extern "C" {
    /// Generate seccomp Berkeley Packet Filter (BPF) code and export it to a buffer
    ///
    /// - `ctx`: the filter context
    /// - `buf`: the destination buffer
    /// - `len`: on input the length of the buffer, on output the number of bytes in the program
    ///
    /// This function generates seccomp Berkeley Packer Filter (BPF) code and writes
    /// it to the given buffer.  Returns zero on success, negative values on failure.
    pub fn seccomp_export_bpf_mem(ctx: const_scmp_filter_ctx, buf: *mut /* FIXME */, len: *mut c_size_t) -> c_int;
}
```

`libseccomp`:
```rust
impl ScmpFilterContext {
    pub fn export_bpf_mem(&self) -> Vec<u8> {
        ...
    }
}
```

@rusty-snake 

I have tried to avoid the failure at link time of the libseccomp using `pkgconfig` and `#![cfg(libseccomp_v2_5)]`.
As you know, our codes of the `notify` feature that requires >=2.5.0 are actually built only when the libseccomp is >=2.50.
This can prevent safely users from using the feature on <2.5.0 and print a good error message.

Example:
```sh
error[E0599]: no method named `get_notify_fd` found for struct `ScmpFilterContext` in the current scope
   --> *************
    |
205 |             ctx.get_notify_fd()
    |                 ^^^^^^^^^^^^^ method not found in `ScmpFilterContext`

For more information about this error, try `rustc --explain E0599`.
```  

Without this feature, the error message will be as follows.
This is a raw and huge error message.
```sh
error: linking with `cc` failed: exit status: 1
*********************
a lot of information
*********************
..............
          ***notify.rs:89: undefined reference to `seccomp_notify_fd'
          collect2: error: ld returned 1 exit status

  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
```

In addition,  `pkgconfig` and `#![cfg(libseccomp_v2_5)]` help our test run properly on <2.5.0.
On <2.5.0, the test codes related to the notify are not built and the other tests run properly without any problem.
Without it, the test will fail at link time due to `undefined reference` like the above message.

On the other hand, I really have wanted to remove the dependency of `pkgconfig` in order to improve maintainability.
I found the `pkgconfig` brought about a problem while I was preparing for the integration test of the libseccomp on https://github.com/ManaSugi/libseccomp/commit/e60406319712477af63a0cb4a85c8811c71efb9c . To build our libseccomp-rs on the CI environment of the libseccomp, I have to create a `pkgconfig` file in advance and it is not beautiful.

Anyway, I like to make the libseccomp-rs fail at link time safely (early detection) because the weak symbol can bring about a failure at runtime (late detection). So, it's good time to reconsider the solution without `pkgconfig`.
About our unit tests, we have to make the new functions not compiled with some features or other.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   You can however create and export seccomp-bpf for them.
 - `SeccompError::sysrawrc()` that queries the system's raw error code directly returned
   by `ScmpFilterAttr::ApiSysRawRc`.
+- `ScmpFilterContext::export_bpf_mem`
 
 ### Changed
 

--- a/libseccomp/src/filter_context.rs
+++ b/libseccomp/src/filter_context.rs
@@ -1021,7 +1021,7 @@ impl ScmpFilterContext {
     ///
     /// # Errors
     ///
-    /// If this function is called with an invalid filter or  writing to the file fails,
+    /// If this function is called with an invalid filter or writing to the file fails,
     /// an error will be returned.
     ///
     /// # Examples
@@ -1049,7 +1049,7 @@ impl ScmpFilterContext {
     ///
     /// # Errors
     ///
-    /// If this function is called with an invalid filter or  writing to the file fails,
+    /// If this function is called with an invalid filter or writing to the file fails,
     /// an error will be returned.
     ///
     /// # Examples
@@ -1101,6 +1101,49 @@ impl ScmpFilterContext {
     #[must_use]
     pub fn as_ptr(&self) -> scmp_filter_ctx {
         self.ctx.as_ptr()
+    }
+}
+
+/// This `impl`-block requires libseccomp 2.6.0 or newer.
+#[cfg(any(libseccomp_v2_6, all(doc, not(doctest))))]
+impl ScmpFilterContext {
+    /// Outputs BPF(Berkeley Packet Filter)-formatted, kernel-readable dump of a
+    /// filter context's rules to a in-memory buffer.
+    ///
+    /// This function corresponds to
+    /// [`seccomp_export_bpf_mem`](https://man7.org/linux/man-pages/man3/seccomp_export_bpf_mem.3.html).
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters an issue while exporting the filter, an error will be returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use libseccomp::*;
+    /// # use std::io::{stdout};
+    /// let ctx = ScmpFilterContext::new(ScmpAction::Allow)?;
+    /// let buf = ctx.export_bpf_mem()?;
+    /// println!("{buf:?}");
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    pub fn export_bpf_mem(&self) -> Result<Vec<u8>> {
+        // We call seccomp_export_bpf_mem with buf specified as a NULL-ptr first
+        // to query the required buffer size.
+        let mut len: usize = 0;
+        cvt(unsafe { seccomp_export_bpf_mem(self.ctx.as_ptr(), std::ptr::null_mut(), &mut len) })?;
+
+        let mut buf: Vec<u8> = vec![0; len];
+        let mut buf_len = buf.len();
+        cvt(unsafe {
+            seccomp_export_bpf_mem(
+                self.ctx.as_ptr(),
+                buf.as_mut_ptr().cast::<libc::c_void>(),
+                &mut buf_len,
+            )
+        })?;
+
+        Ok(buf)
     }
 }
 

--- a/libseccomp/tests/tests.rs
+++ b/libseccomp/tests/tests.rs
@@ -332,11 +332,19 @@ fn test_export_functions() {
 
     let mut invalid_fd: File = unsafe { std::fs::File::from_raw_fd(-2) };
 
+    assert!(ctx.export_pfc(&mut stdout()).is_ok());
+    assert!(ctx.export_pfc(&mut invalid_fd).is_err());
+
     assert!(ctx.export_bpf(&mut stdout()).is_ok());
     assert!(ctx.export_bpf(&mut invalid_fd).is_err());
 
-    assert!(ctx.export_pfc(&mut stdout()).is_ok());
-    assert!(ctx.export_pfc(&mut invalid_fd).is_err());
+    #[cfg(libseccomp_v2_6)]
+    {
+        let ctx = ScmpFilterContext::new(ScmpAction::Allow).unwrap();
+        let buf = ctx.export_bpf_mem().unwrap();
+        assert!(buf.len() > 0);
+        assert!(buf.iter().copied().any(|byte| byte > 0));
+    }
 }
 
 #[test]

